### PR TITLE
Fix and unify channel and user list items

### PR DESCRIPTION
**Pull request by larkox**
#### Summary
This PR tries to unify all the channel and user list items. This unify how they are presented, and fix some issues like the linked ticket.

Tested on Android, iOS and iPad. Not sure if everything is working as it used to do, but seems like everything (or most of it) is working and looks fine. 0/5 on whether we should send it to QA or UX, since things may have changed around.

Disclaimer:
There is another refactor I would like to do to simplify the items (mainly the channel item) that has to do with the background color. Right now, there is a lot of logic that before was about "isInfo" but that I renamed to "onCenterBg", which means that the element is on a white background (on denim theme) and the texts have to be related to that. The refactor I want to do is to handle this in a more beautiful way (probably inside the UserItem/ChannelItem component) instead of where it is now.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-45498

#### Release Note
```release-note
Improve channel and user elements in lists.
```

### CODE REVIEW

**Comment by larkox** [comment]
Updated the description with:
Tested on Android, iOS and iPad. Not sure if everything is working as it used to do, but seems like everything (or most of it) is working and looks fine. 0/5 on whether we should send it to QA or UX, since things may have changed around.

**Comment by M-ZubairAhmed** [comment]
@larkox can you check the conflicts?

**Comment by larkox** [comment]
@M-ZubairAhmed done

File: app/components/user_item/user_item.tsx
Comment (enahum) [inline app/components/user_item/user_item.tsx:39]: is this really needed?

**Review by enahum — COMMENTED** [review]
There are multiple Style and view changes, can you please share screenshots of before and after of all affected areas ?

File: app/components/autocomplete/at_mention_item/index.tsx
Comment (larkox) [reply app/components/autocomplete/at_mention_item/index.tsx:23]: Because it is not. The username comes from the arguments, not from the prop.

File: app/components/channel_item/channel_item.tsx
Comment (larkox) [reply app/components/channel_item/channel_item.tsx:73]: The default badge style, among other things, set the position to absolute, and a few other things that we don't want in this context. I didn't want to change that in the badge component, to make sure I didn't break other places where it was working right. That being said, maybe it make sense to change it. I will look a bit deeper into it.

**Comment by enahum** [comment]
conflicts need resolving and I'll wait for this to be merged in order to work on https://mattermost.atlassian.net/browse/MM-49383 as currently the snackbar is covered by the selected users panel and I don't want to be doing more than actually needed.

**Comment by larkox** [comment]
Adding screenshots:

Android: Practically the same as iPhone.

iPhone:
Simulator Screen Shot - iPhone 13 - 2023-04-03 at 11 36 09
Simulator Screen Shot - iPhone 13 - 2023-04-03 at 11 36 50
Simulator Screen Shot - iPhone 13 - 2023-04-03 at 11 37 09
Simulator Screen Shot - iPhone 13 - 2023-04-03 at 11 37 19
Simulator Screen Shot - iPhone 13 - 2023-04-03 at 11 37 33
Simulator Screen Shot - iPhone 13 - 2023-04-03 at 11 37 46
Simulator Screen Shot - iPhone 13 - 2023-04-03 at 11 37 58
Simulator Screen Shot - iPhone 13 - 2023-04-03 at 11 38 18
Simulator Screen Shot - iPhone 13 - 2023-04-03 at 11 38 32
Simulator Screen Shot - iPhone 13 - 2023-04-03 at 11 38 50
Simulator Screen Shot - iPhone 13 - 2023-04-03 at 11 38 54


iPad:
Simulator Screen Shot - iPad Pro (12 9-inch) (5th generation) - 2023-04-03 at 11 39 18
Simulator Screen Shot - iPad Pro (12 9-inch) (5th generation) - 2023-04-03 at 11 39 28
Simulator Screen Shot - iPad Pro (12 9-inch) (5th generation) - 2023-04-03 at 11 39 36
Simulator Screen Shot - iPad Pro (12 9-inch) (5th generation) - 2023-04-03 at 11 39 49
Simulator Screen Shot - iPad Pro (12 9-inch) (5th generation) - 2023-04-03 at 11 40 04
Simulator Screen Shot - iPad Pro (12 9-inch) (5th generation) - 2023-04-03 at 11 40 12
Simulator Screen Shot - iPad Pro (12 9-inch) (5th generation) - 2023-04-03 at 11 40 54
Simulator Screen Shot - iPad Pro (12 9-inch) (5th generation) - 2023-04-03 at 11 41 08
Simulator Screen Shot - iPad Pro (12 9-inch) (5th generation) - 2023-04-03 at 11 41 17
Simulator Screen Shot - iPad Pro (12 9-inch) (5th generation) - 2023-04-03 at 11 41 32
Simulator Screen Shot - iPad Pro (12 9-inch) (5th generation) - 2023-04-03 at 11 41 41
Simulator Screen Shot - iPad Pro (12 9-inch) (5th generation) - 2023-04-03 at 11 41 45

File: app/components/channel_icon/index.tsx
Comment (enahum) [inline app/components/channel_icon/index.tsx:224]: does it need the fragment?

```suggestion
    return icon;
```

File: app/components/selected_chip/index.tsx
Comment (enahum) [inline app/components/selected_chip/index.tsx:78]: why the replace?

**Review by enahum — APPROVED** [review]
LGTM

File: app/components/selected_chip/index.tsx
Comment (larkox) [reply app/components/selected_chip/index.tsx:78]: Funny story. Imagine the text `SomeLongWordHere Short`. Now imagine that the space we have cannot contain `SomeLongWordHere`. So it has to calculate how much space it can show. Then for some reason... that calculation, instead of being the actual available space, it is the length of `Short`. So... instead of having something like `SomeLongWor...` you end up with `SomeL...`.

Doing this replace, we are changing normal spaces by non-breaking spaces. That way, it cannot separate the words to do this calculation, and does the calculation right.

No idea why it does this, and it got me way too long to figure out this workaround, but now it seems to work as it should.

File: app/components/channel_icon/index.tsx
Comment (larkox) [reply app/components/channel_icon/index.tsx:224]: It was complaining because `icon` can be undefined. If we set icon to `null` instead, typescript doesn't yell.

File: app/components/selected_chip/index.tsx
Comment (enahum) [reply app/components/selected_chip/index.tsx:78]: This is an incredible find, we should share this more broadly but have you looked into the break strategy props for ios and android? https://reactnative.dev/docs/text#textbreakstrategy-android

0/5 if we want to change it, but we should at least give it a try

Comment (larkox) [reply app/components/selected_chip/index.tsx:78]: Setting the textbreakstrategy to 'simple', works better for `SomeLongWordHere Short`, but not so well for `Short SomeLongWordHere` :sweat_smile:

`balanced` seems to solve the issue in most scenarios, but not all of them. I still cannot figure out exactly why. The most consistent solution seems to be the non-breaking spaces.

And even if this was consistent, that would solve for API +24, and I think we support a few behind that, right?

I just added this approach for the `user_item`, where this was also giving some minor trouble, and also to other composed strings. For example (simplified):
```
<Text style={styles.normal}>
   {displayName}
    <Text style={style.dimmed}>
        {` ~${channelName}`}
    </Text>
</Text>
```

Comment (enahum) [reply app/components/selected_chip/index.tsx]: Thanks for trying

**Review by enahum — APPROVED** [review]

**Comment by larkox** [comment]
@M-ZubairAhmed Not urgent, but it would be great if you could review this one. Kind of big one, so I would like to merge it before it becomes a merge hell :sweat_smile:

**Comment by enahum** [comment]
> @M-ZubairAhmed Not urgent, but it would be great if you could review this one. Kind of big one, so I would like to merge it before it becomes a merge hell 😅

also keep in mind that I have a couple of tickets waiting for this to be merged ;)

File: app/components/channel_icon/index.tsx
Comment (M-ZubairAhmed) [inline app/components/channel_icon/index.tsx:146]: In webapp we have deprecated use of compass icons should we do the same here as well?

File: app/components/channel_item/channel_body.tsx
Comment (M-ZubairAhmed) [inline app/components/channel_item/channel_body.tsx:24]: This is an interesting concept of styling in mobile

File: app/components/channel_item/index.ts
Comment (M-ZubairAhmed) [inline app/components/channel_item/index.ts:52]: What's of$

File: app/components/user_item/user_item.tsx
Comment (M-ZubairAhmed) [inline app/components/user_item/user_item.tsx:51]: What's the difference between create and other methods to create style

**Review by M-ZubairAhmed — APPROVED** [review]

File: app/components/channel_icon/index.tsx
Comment (enahum) [reply app/components/channel_icon/index.tsx:146]: no, here the SVG version is much much heavier than using the font directly.

File: app/components/user_item/user_item.tsx
Comment (enahum) [reply app/components/user_item/user_item.tsx:51]: `makeStyleSheetFromTheme` uses a memoized version of the `theme` while this does not, but in this case, there are no colors that come from the theme, so there is no need to memoize it.

File: app/components/channel_item/index.ts
Comment (enahum) [reply app/components/channel_item/index.ts:52]: https://rxjs.dev/api/index/function/of

**Pull request merged**
